cmd/archive_upload_mass_upload/mass_upload.go

```diff
@@ -195,42 +212,69 @@ func (c *client) ftpWalk(dir string) {
 	if w.Err() != nil {
 		glog.Errorf("Next returned false, closing channel and returning: %v", w.Err())
 		glog.Errorf("Current working directory: %s", w.Path())
+		// Close the channel, the readChannel threads should continue to drain this channel.
 		close(c.ch)
 		return
 	}
 }
 
+func (c *client) metric(k string) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.metrics[k]++
+}
+
 // readChannel reads FTP file results from a channel, collects and compares MD5 checksums
 // and uploads files to cloud-storage if mismatches occur.
 func (c *client) readChannel(ctx context.Context) {
-	errs := 0
+	defer c.wg.Done()
+	ftpErrs := 0
+	grpcErrs := 0
+
+	// Open a new, bespoke FTP connection, so overlapping
+	// command/data channel problems are avoided.
+	f, err := connectFtp(c.site)
+	if err != nil {
+		glog.Errorf("failed to open new FTP connection: %v", err)
+		return
+	}
+	if err := f.Login(c.user, c.passwd); err != nil {
+		glog.Errorf("failed to login to FTP site: %v", err)
+		return
+	}
+
 	for {
 		ef, ok := <-c.ch
-		// Exit if ef is nil, because the channel closed.
+		// Exit if ok is false, this is the 'channel closed' signal.
 		if !ok {
 			glog.Error("Channel closed, exiting readChannel.")
 			break
 		}
 
 		fn := strings.TrimLeft(ef.name, "/")
+		// If the file doesn't have UPDATES/updates do not process further.
+		if !strings.Contains(fn, "UPDATES/updates") {
+			continue
+		}
```

Comment:
I suppose we may need this tool for RIPE as well...? I don't think RIPE paths have "UPDATES/updates".

Reply:
Good point.. I imagine a map of:
  proto.DATASOURCE -> important-path-attribute

is in our future :) Ripe appears to store in this central location:
  https://data.ris.ripe.net/rrc11/2021.03/updates.20210331.225.gz
  
  rrc11 == 1 collector name... there are ~30-odd collectors.
Before we start digesting that data we should chat up their lead for these things, who I think is:
   Robert Kisteleki (he's easy to get ahold of)